pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -626,6 +626,11 @@ func (s *DiscoveryServer) initConnectionNode(discReq *xdsapi.DiscoveryRequest, c
 	if err := nt.SetServiceInstances(s.Env); err != nil {
 		return err
 	}
+	// If the proxy has no service instances and its a gateway, kill the XDS connection as we cannot
+	// serve any gateway config if we dont know the proxy's service instances
+	if nt.Type == model.Router && (nt.ServiceInstances == nil || len(nt.ServiceInstances) == 0) {
+		return errors.New("gateway has no associated service instances")
```

Comment:
nit: should the node.ID field be included in the error message?  
also this error is returned directly in ads.go and never logged anywhere in pilot.  The error appears in envoy's logs for the gateway instance in question but doesn't provide any context to the error in pilot

Reply:
yes.. I saw that.. open to suggestions on fixing this